File: Mesh4D/Matlab4D/Tensor/examples/example_concave.m
clear;
clc;
close all;

projectRoot = fileparts(fileparts(mfilename('fullpath')));
addpath(projectRoot);
addpath(fullfile(projectRoot, 'src'));

%**************************************************************************
%
%   Esempio:
%       Cubatura 4D su dominio poliedrale concavo in movimento
%
%   Descrizione:
%       Questo esempio dimostra l'utilizzo del metodo
%       OPC4D_Tensor per l'integrazione di una funzione su un
%       dominio poliedrale concavo in movimento per tau in [0,1],
%       rappresentato mediante una mesh superficiale triangolare.
%
%       Il test considera la funzione costante f = 1. L'integrale 4D esatto
%       viene calcolato integrando analiticamente il volume istantaneo del
%       dominio lungo l'intervallo temporale [0,1].
%
%**************************************************************************

%**************************************************************************
% Parametri e caricamento mesh
%**************************************************************************

ade = 1;

fprintf('\n');
fprintf('**************************************************************\n');
fprintf('                   OPC4D_Tensor\n');
fprintf('              - Metodo Tensoriale -\n');
fprintf('    Cubatura su Dominio Poliedrale Concavo in Movimento\n');
fprintf('**************************************************************\n');
fprintf('\n');

fprintf('Ade:                   %d\n', ade);

vertici_iniziali = load('concave_vertex.dat');
facets           = load('concave_tri.dat');
vertici_finali   = load('concave_vertex_new.dat');

fprintf('Numero di vertici:     %d\n', size(vertici_iniziali,1));
fprintf('Numero di facce:       %d\n', size(facets,1));

%**************************************************************************
% Definizione funzione integranda
%**************************************************************************

f = @(x,y,z,tau) ones(size(x));
f_string = 'f(x,y,z,tau) = 1';

fprintf('Funzione integranda:   %s\n', f_string);

%**************************************************************************
% Plot dominio poliedrale iniziale e finale
%**************************************************************************

figure('Color','w');

h = patch('Vertices', vertici_iniziali, ...
          'Faces', facets, ...
          'FaceColor', [0.7 0.7 0.7], ...
          'EdgeColor', [0.2 0.2 0.2], ...
          'FaceAlpha', 0.65);

axis equal;
grid on;
box on;

xlabel('$x$', 'Interpreter','latex');
ylabel('$y$', 'Interpreter','latex');
zlabel('$z$', 'Interpreter','latex');

view(3);

% Limiti degli assi fissati sull'intero movimento
vertici_complessivi = [vertici_iniziali; vertici_finali];

xlim([min(vertici_complessivi(:,1)), max(vertici_complessivi(:,1))]);
ylim([min(vertici_complessivi(:,2)), max(vertici_complessivi(:,2))]);
zlim([min(vertici_complessivi(:,3)), max(vertici_complessivi(:,3))]);

%**************************************************************************
% Animazione del dominio
%**************************************************************************

n_frame = 100;

for k = 1:n_frame

    tau = (k-1)/(n_frame-1);

    % Interpolazione convessa tra configurazione iniziale e finale
    vertici_tau = (1-tau) * vertici_iniziali + ...
                  tau * vertici_finali;

    % Aggiornamento della geometria
    set(h, 'Vertices', vertici_tau);

    % Aggiornamento del titolo
    title(sprintf( ...
        'Dominio Poliedrale Concavo in Movimento, $\\tau = %.2f$', tau), ...
        'Interpreter','latex');

    drawnow;

    % Controllo della velocita' dell'animazione
    pause(0.01);

end


%**************************************************************************
% Inizio regola di cubatura
%**************************************************************************

fprintf('\n');
fprintf('Inizio Cubatura...\n');

tic;
[XYZT, W] = OPC4D_Tensor(ade, vertici_iniziali,vertici_finali,facets, 'GJCC');
% [XYZT, W] = OPC4D_Tensor(ade, vertici_iniziali,vertici_finali,facets, 'GJL');
% [XYZT, W] = OPC4D_Tensor(ade, vertici_iniziali,vertici_finali,facets, 'DGL');
% [XYZT, W] = OPC4D_Tensor(ade, vertici_iniziali,vertici_finali,facets, 'DCC');
elapsedTime = toc;

I = W' * f(XYZT(:,1), XYZT(:,2), XYZT(:,3), XYZT(:,4));


%**************************************************************************
% Visualizzazione risultati e punti di cubatura
%**************************************************************************

fprintf('Fine Cubatura...\n');
fprintf('\n');

fprintf('Numero di nodi 4D:    %d\n', size(XYZT,1));
fprintf('Integrale numerico:   %.15e\n', I);
fprintf('Tempo di calcolo:     %.6e s\n', elapsedTime);

File: Mesh4D/Fortran4D/Tensor/examples/example_concave.f90
PROGRAM example_convex

    USE TypesDef, ONLY: dp
    USE PolyhedronMesh, ONLY: t_polyhedron, MeshReader
    USE OPC4D_Tensor_Module, ONLY: OPC4D_Tensor

    IMPLICIT NONE

    !***********************************************************************
    !
    !   Esempio:
    !       Cubatura 4D su dominio poliedrale convesso in movimento
    !
    !   Descrizione:
    !       Questo esempio dimostra l'utilizzo del metodo
    !       OptimalPolyCuba4D Tensor per l'integrazione di una funzione su
    !       un dominio spazio-temporale (x, y, z, tau), ottenuto dalla
    !       deformazione lineare, per tau in [0,1], di un poliedro convesso
    !       rappresentato da una mesh superficiale triangolare.
    !       La configurazione iniziale (tau = 0) e quella finale (tau = 1)
    !       sono lette rispettivamente da convex_vertex.dat e
    !       convex_vertex_new.dat, con la stessa connettivita' delle facce.
    !
    !***********************************************************************

    !***********************************************************************
    ! Dichiarazione delle variabili
    !***********************************************************************

    INTEGER                         :: ade
    INTEGER                         :: n_vertici, n_facce
    REAL(dp)                        :: Integrale
    INTEGER                         :: i
    REAL(dp), ALLOCATABLE           :: XYZT(:,:)
    REAL(dp), ALLOCATABLE           :: W(:)
    REAL(dp)                        :: elapsedTime
    REAL(dp)                        :: t_start, t_end

    TYPE(t_polyhedron)              :: poly

    CHARACTER(LEN=256)              :: vertici_file
    CHARACTER(LEN=256)              :: vertici_new_file
    CHARACTER(LEN=256)              :: tri_file

    !***********************************************************************
    ! Parametri e caricamento mesh
    !***********************************************************************

    ade = 10

    WRITE(*,'(A)')
    WRITE(*,'(A)') '**************************************************************'
    WRITE(*,'(A)') '                       OPC4D Tensor'
    WRITE(*,'(A)') '    Cubatura su Dominio Poliedrale Convesso in Movimento'
    WRITE(*,'(A)') '**************************************************************'
    WRITE(*,'(A)')

    WRITE(*,'(A,I0)') 'Ade:                   ', ade

    ! Percorsi relativi alla cartella examples.
    vertici_file     = 'concave_vertex.dat'
    vertici_new_file = 'concave_vertex_new.dat'
    tri_file         = 'concave_tri.dat'

    ! Lettura della mesh superficiale triangolare (configurazione iniziale
    ! e finale) e della connettivita' delle facce.
    CALL MeshReader(vertici_file, vertici_new_file, tri_file, poly)

    n_vertici = SIZE(poly%vertici,1)
    n_facce   = SIZE(poly%facce,1)

    WRITE(*,'(A,I0)') 'Numero di vertici:     ', n_vertici
    WRITE(*,'(A,I0)') 'Numero di facce:       ', n_facce

    !***********************************************************************
    ! Definizione funzione integranda
    !***********************************************************************

    WRITE(*,'(A)') 'Funzione integranda:   f(x,y,z,tau) = 1'

    !***********************************************************************
    ! Inizio regola di cubatura
    !***********************************************************************

    WRITE(*,'(A)')
    WRITE(*,'(A)') 'Inizio Cubatura...'

    CALL CPU_TIME(t_start)

    CALL OPC4D_Tensor(ade, poly%vertici, poly%vertici_new, poly%facce, 'GJCC', XYZT, W)
    ! CALL OPC4D_Tensor(ade, poly%vertici, poly%vertici_new, poly%facce, 'GJL', XYZT, W)
    ! CALL OPC4D_Tensor(ade, poly%vertici, poly%vertici_new, poly%facce, 'DCC', XYZT, W)
    ! CALL OPC4D_Tensor(ade, poly%vertici, poly%vertici_new, poly%facce, 'DGL', XYZT, W)

    CALL CPU_TIME(t_end)

    elapsedTime = t_end - t_start

    Integrale = 0.0_dp
    DO i = 1, SIZE(W)
        Integrale = Integrale + W(i) * f(XYZT(i,1), XYZT(i,2), XYZT(i,3), XYZT(i,4))
    END DO

    !***********************************************************************
    ! Visualizzazione risultati e punti di cubatura
    !***********************************************************************

    WRITE(*,'(A)') 'Fine Cubatura...'

    WRITE(*,'(A)')
    WRITE(*,'(A,I0)') 'Numero di nodi 4D:      ', SIZE(W)
    WRITE(*,'(A,F0.15)') 'Integrale numerico:       ', Integrale
    WRITE(*,'(A,ES14.6,A)') 'Tempo di calcolo:       ', elapsedTime, ' s'

    CONTAINS

    !***********************************************************************
    ! Funzione integranda
    !***********************************************************************

    REAL(dp) FUNCTION f(x,y,z,tau)

        IMPLICIT NONE

        REAL(dp), INTENT(IN) :: x
        REAL(dp), INTENT(IN) :: y
        REAL(dp), INTENT(IN) :: z
        REAL(dp), INTENT(IN) :: tau

        f = 1.0_dp

    END FUNCTION f

END PROGRAM example_convex